File: hydra-haskell/src/main/haskell/Hydra/Dsl/Terms.hs
-- | A domain-specific language for constructing Hydra terms in Haskell.
module Hydra.Dsl.Terms where

import Hydra.Constants
import Hydra.Core
import Hydra.Graph
import Hydra.Mantle
import Hydra.Dsl.Common
import qualified Hydra.Dsl.Literals as Literals

import Prelude hiding (map, product, sum)
import Data.Int
import qualified Data.Char as C
import qualified Data.List as L
import qualified Data.Map as M
import qualified Data.Set as S


-- Operators

-- | Function composition operator: f <.> g creates a function that applies g then f
-- Example: var "stringLength" <.> var "toString"
(<.>) :: Term -> Term -> Term
f <.> g = compose f g

-- | Term application operator: function @@ argument
-- Example: var "add" @@ int32 1
(@@) :: Term -> Term -> Term
fun @@ arg = apply fun arg

-- | Field definition operator: name>: value
-- Example: "name">: string "John"
infixr 0 >:
(>:) :: String -> Term -> Field
name>: term = field name term


-- | Attach an annotation to a term
-- Example: annot (M.fromList [(Name "comment", string "A User ID")]) (var "userId")
annot :: M.Map Name Term -> Term -> Term
annot ann term = TermAnnotated $ AnnotatedTerm term ann

-- | Attach an annotation to a term
-- Example: annotated (var "userId") (M.fromList [(Name "comment", string "A User ID")])
annotated :: Term -> M.Map Name Term -> Term
annotated term ann = TermAnnotated $ AnnotatedTerm term ann

-- | Apply a function term to an argument
-- Example: apply (var "capitalize") (string "arthur")
apply :: Term -> Term -> Term
apply fun arg = TermApplication $ Application fun arg

applyAll :: Term -> [Term] -> Term
applyAll fun args = foldl apply fun args

-- | Create a binary data literal. The encoding scheme is currently application-dependent.
-- Example: binary ""\x48\x65\x00\xff\x20\x7a\x1b\x80"
binary :: String -> Term
binary = literal . Literals.binary

-- | Create a bigfloat literal. Note: in practice, precision is limited to 64 bits (same as Double) in Haskell.
-- Example: bigfloat 3.14159265359
bigfloat :: Double -> Term
bigfloat = literal . Literals.bigfloat

-- | Create a bigint literal
-- Example: bigint 9223372036854775808
bigint :: Integer -> Term
bigint = literal . Literals.bigint

-- | Create a boolean literal
-- Example: boolean True
boolean :: Bool -> Term
boolean = literal . Literals.boolean

char :: Char -> Term
char = int32 . C.ord

comparison :: Comparison -> Term
comparison t = case t of
  ComparisonEqualTo -> unitVariant _Comparison _Comparison_equalTo
  ComparisonLessThan -> unitVariant _Comparison _Comparison_lessThan
  ComparisonGreaterThan -> unitVariant _Comparison _Comparison_greaterThan

-- | Compose two functions (apply g then f) to create a new function
-- Example: compose (var "stringLength") (var "toString")
-- This creates a function equivalent to \x -> stringLength(toString(x))
-- Function composition applies right-to-left: (f ∘ g)(x) = f(g(x))
compose :: Term -> Term -> Term
compose f g = lambda "arg_" $ apply f (apply g $ var "arg_")

-- | Create a constant function that always returns the same value
-- Example: constant true
constant :: Term -> Term
constant = lambda ignoredVariable

-- | Boolean false literal
false :: Term
false = boolean False

-- | Create a field with the given name and value
-- Example: field "age" (int32 30)
field :: String -> Term -> Field
field n = Field (Name n)

-- | First element projection function for pairs
first :: Term
first = untuple 2 0

-- | Create a floating-point literal with specified precision
-- Example: float (FloatValueFloat32 3.14)
float :: FloatValue -> Term
float = literal . Literals.float

-- | Create a float32 literal
-- Example: float32 3.14
float32 :: Float -> Term
float32 = literal . Literals.float32

-- | Create a float64 literal
-- Example: float64 3.14159265359
float64 :: Double -> Term
float64 = literal . Literals.float64

-- | Identity function
identity :: Term
identity = lambda "x_" $ var "x_"

-- | Create a union value by injecting a value into a specific variant
-- Example: inject (Name "Result") ("success">: int32 42)
-- This creates a "Result" union with the "success" variant containing value 42
-- Use this to construct values of union types at runtime
inject :: Name -> Field -> Term
inject tname = TermUnion . Injection tname

-- | Create an int8 literal
-- Example: int8 127
int8 :: Int8 -> Term
int8 = literal . Literals.int8

-- | Create an int16 literal
-- Example: int16 32767
int16 :: Int16 -> Term
int16 = literal . Literals.int16

-- | Create an int32 literal
-- Example: int32 42
int32 :: Int -> Term
int32 = literal . Literals.int32

-- | Create an int64 literal
-- Example: int64 9223372036854775807
int64 :: Int64 -> Term
int64 = literal . Literals.int64

-- | Create an integer literal with specified bit width
-- Example: integer (IntegerValueInt32 42)
integer :: IntegerValue -> Term
integer = literal . Literals.integer

-- | Create a 'Just' optional value
-- Example: just (string "found")
just :: Term -> Term
just = optional . Just

-- | Create a lambda function with one parameter
-- Example: lambda "x" (var "x" @@ int32 1)
lambda :: String -> Term -> Term
lambda param body = TermFunction $ FunctionLambda $ Lambda (Name param) Nothing body

-- | Create a multi-parameter lambda function (curried)
-- Example: lambdas ["x", "y"] (var "add" @@ var "x" @@ var "y")
-- This creates the function \x.\y.add x y
lambdas :: [String] -> Term -> Term
lambdas params body = case params of
  [] -> body
  (h:r) -> lambda h $ lambdas r body

-- | Create a lambda function with a given domain
-- Example: lambdaTyped "x" Types.int32 (list [var "x"])
lambdaTyped :: String -> Type -> Term -> Term
lambdaTyped param dom body = TermFunction $ FunctionLambda $ Lambda (Name param) (Just dom) body

-- | Create a let term with any number of bindings
-- Example: lets ["x">: int32 1, "y">: int32 2] (pair (var "x") (var "y"))
lets :: [Field] -> Term -> Term
lets bindings env = TermLet $ Let (toBinding <$> bindings) env
  where
    toBinding (Field name value) = Binding name value Nothing

letsTyped :: [(String, Term, TypeScheme)] -> Term -> Term
letsTyped bindings env = TermLet $ Let (toBinding <$> bindings) env
  where
    toBinding (name, value, ts) = Binding (Name name) value (Just ts)

-- | Create a list of terms
-- Example: list [int32 1, int32 2, int32 3]
list :: [Term] -> Term
list = TermList

-- | Create a term from a literal value
-- Example: literal (LiteralString "hello")
literal :: Literal -> Term
literal = TermLiteral

-- | Create a map/dictionary term
-- Example: map (M.fromList [(string "January", int32 31), (string "February", int32 28)])
map :: M.Map Term Term -> Term
map = TermMap

-- | Create a pattern match on a union type
-- Example: match (Name "Result") (Just (string "unknown"))
--               ["success">: lambda "s" (var "processSuccess" @@ var "s"),
--                "error">: lambda "e" (var "handleError" @@ var "e")]
-- This allows handling different cases of a union type with specific logic for each variant.
-- The optional second parameter provides a default case for any unmatched variants.
match :: Name -> Maybe Term -> [Field] -> Term
match tname def fields = TermFunction $ FunctionElimination $ EliminationUnion $ CaseStatement tname def fields

-- | Create a pattern match using variant name pairs
-- Example: matchWithVariants (Name "Result") Nothing [(Name "success", Name "handleSuccess"), (Name "error", Name "handleError")]
matchWithVariants :: Name -> Maybe Term -> [(Name, Name)] -> Term
matchWithVariants tname def pairs = match tname def (toField <$> pairs)
  where
    toField (from, to) = Field from $ constant $ unitVariant tname to

-- | Create a 'Nothing' optional value
nothing :: Term
nothing = optional Nothing

-- | Create an optional (nullable) term
-- Example: optional (Just (string "found"))
optional :: Maybe Term -> Term
optional = TermMaybe

-- | Create a pair (2-tuple)
-- Example: pair (string "name") (int32 42)
pair :: Term -> Term -> Term
pair a b = TermProduct [a, b]

-- | Create a primitive function
-- Example: primitive (Name "hydra.lib.strings.length")
primitive :: Name -> Term
primitive = TermFunction . FunctionPrimitive

-- | Create a field projection function
-- Example: project (Name "Person") (Name "firstName")
project :: Name -> Name -> Term
project tname fname = TermFunction $ FunctionElimination $ EliminationRecord $ Projection tname fname

-- | Create a record with named fields of the specified type
-- Example: record (Name "Person") [
--            "name">: string "John",
--            "age">: int32 30,
--            "email">: string "john@example.com"]
-- Records are products of named fields with values that can be accessed by field name
record :: Name -> [Field] -> Term
record tname fields = TermRecord $ Record tname fields

-- | Second element projection function for pairs
second :: Term
second = untuple 2 1

-- | Create a set of terms
-- Example: set (S.fromList [string "a", string "b", string "c"])
set :: S.Set Term -> Term
set = TermSet

-- | Create a string literal
-- Example: string "hello world"
string :: String -> Term
string = TermLiteral . LiteralString

-- | Create a sum term
-- Example: sum 0 3 (int32 1) represents the first element of a 3-element sum
sum :: Int -> Int -> Term -> Term
sum idx arity term = TermSum $ Sum idx arity term

triple :: Term -> Term -> Term -> Term
triple a b c = tuple [a, b, c]

-- | Boolean true literal
true :: Term
true = boolean True

-- | Create a product (tuple) with multiple components
-- Example: product [string "name", int32 42, true]
tuple :: [Term] -> Term
tuple = TermProduct

tuple4 :: Term -> Term -> Term -> Term -> Term
tuple4 a b c d = tuple [a, b, c, d]

tuple5 :: Term -> Term -> Term -> Term -> Term -> Term
tuple5 a b c d e = tuple [a, b, c, d, e]

tyapp :: Term -> Type -> Term
tyapp term typ = TermTypeApplication $ TypeApplicationTerm term typ

tyapps = typeApplication

tylam :: String -> Term -> Term
tylam var body = TermTypeLambda $ TypeLambda (Name var) body

tylams :: [String] -> Term -> Term
tylams vars body = L.foldl (\b v -> TermTypeLambda $ TypeLambda (Name v) b) body $ L.reverse vars

-- | Create a type abstraction (universal quantification)
-- Example: typeLambda [Name "a", Name "b"] (lambdaTyped "f" (Types.function (Types.var "a") (Types.var "b"))
--                                               (lambdaTyped "x" (Types.var "a") (var "f" @@ var "x")))
-- This creates a polymorphic term with type variables.
-- The example creates a higher-order function with type 'forall a b. (a -> b) -> a -> b',
-- which is the polymorphic apply function that works for any types a and b.
typeLambda :: [Name] -> Term -> Term
typeLambda vars body = L.foldl (\b v -> TermTypeLambda $ TypeLambda v b) body vars

-- | Apply type arguments to a polymorphic term
-- Example: typeApplication (var "map") [Types.int32, Types.string]
-- This instantiates a polymorphic function with concrete types.
-- For instance, if 'map' has type 'forall a b. (a -> b) -> list a -> list b',
-- the example would instantiate it to '(int32 -> string) -> list int32 -> list string'.
typeApplication :: Term -> [Type] -> Term
typeApplication term types = L.foldl (\t ty -> TermTypeApplication $ TypeApplicationTerm t ty) term types

-- | Create a uint8 literal
-- Example: uint8 255
uint8 :: Int16 -> Term
uint8 = literal . Literals.uint8

-- | Create a uint16 literal
-- Example: uint16 65535
uint16 :: Int -> Term
uint16 = literal . Literals.uint16

-- | Create a uint32 literal
-- Example: uint32 4294967295
uint32 :: Int64 -> Term
uint32 = literal . Literals.uint32

-- | Create a uint64 literal
-- Example: uint64 18446744073709551615
uint64 :: Integer -> Term
uint64 = literal . Literals.uint64

-- | Unit value (empty record)
unit :: Term
unit = TermUnit

-- | Create a unit variant of a union
-- Example: unitVariant (Name "Result") (Name "success")
unitVariant :: Name -> Name -> Term
unitVariant tname fname = variant tname fname unit

-- | Create a untyped tuple projection function
-- Example: untuple 3 1 Nothing extracts the second element of a 3-tuple
untuple :: Int -> Int -> Term
untuple arity idx = TermFunction $ FunctionElimination $ EliminationProduct $ TupleProjection arity idx Nothing

-- | Create an unwrap function for a wrapped type
-- Example: unwrap (Name "Email")
unwrap :: Name -> Term
unwrap = TermFunction . FunctionElimination . EliminationWrap

-- | Create a variable reference
-- Example: var "x"
var :: String -> Term
var = TermVariable . Name

-- | Create a union variant
-- Example: variant (Name "Result") (Name "success") (string "ok")
variant :: Name -> Name -> Term -> Term
variant tname fname term = TermUnion $ Injection tname $ Field fname term

-- | Create a constant function that produces a unit variant
-- Example: withVariant (Name "Result") (Name "success")
withVariant :: Name -> Name -> Term
withVariant tname = constant . unitVariant tname

-- | Create a wrapped term
-- Example: wrap (Name "Email") (string "user@example.com")
wrap :: Name -> Term -> Term
wrap name term = TermWrap $ WrappedTerm name term

File: hydra-python/src/main/python/hydra/dsl/terms.py
"""A DSL for constructing Hydra terms in Python."""

from collections.abc import Mapping, Sequence
from functools import reduce

import hydra.constants
import hydra.dsl.literals as lt
import hydra.util
from hydra.core import (
    AnnotatedTerm,
    Application,
    CaseStatement,
    Elimination,
    EliminationRecord,
    EliminationUnion,
    EliminationWrap,
    Field,
    FloatValue,
    FunctionElimination,
    FunctionLambda,
    FunctionPrimitive,
    Injection,
    IntegerValue,
    Lambda,
    Let,
    Binding,
    Literal,
    Name,
    Projection,
    Record,
    Term,
    TermAnnotated,
    TermApplication,
    TermEither,
    TermFunction,
    TermLet,
    TermList,
    TermLiteral,
    TermMap,
    TermMaybe,
    TermPair,
    TermRecord,
    TermSet,
    TermTypeApplication,
    TermTypeLambda,
    TermUnion,
    TermUnit,
    TermVariable,
    TermWrap,
    Type,
    TypeApplicationTerm,
    TypeLambda,
    TypeScheme,
    WrappedTerm,
)
from hydra.dsl.python import FrozenDict, Maybe, Just, Nothing, Left, Right


# Operators - TODO: find Python equivalents for these special syntax forms
# (<.>) :: Term -> Term -> Term  -- function composition operator
# (@@) :: Term -> Term -> Term   -- term application operator
# (>:) :: String -> Term -> Field  -- field definition operator


def annot(ann: Mapping[str, Term], t: Term) -> Term:
    """Attach an annotation to a term.

    Example: annot({"comment": string("A User ID")}, var("userId"))
    """
    return TermAnnotated(
        AnnotatedTerm(t, FrozenDict({Name(k): v for k, v in ann.items()}))
    )


def annotated(t: Term, ann: Mapping[str, Term]) -> Term:
    """Attach an annotation to a term (alternative argument order).

    Example: annotated(var("userId"), {"comment": string("A User ID")})
    """
    return annot(ann, t)


def apply(func: Term, arg: Term) -> Term:
    """Apply a function term to an argument.

    Example: apply(var("capitalize"), string("arthur"))
    """
    return TermApplication(Application(func, arg))


def apply_all(func: Term, args: Sequence[Term]) -> Term:
    """Apply a function to multiple arguments (curried application).

    Example: apply_all(var("add"), [int32(1), int32(2)])
    """
    return reduce(apply, args, func)


def bigfloat(value: float) -> Term:
    """Create a bigfloat literal.

    Example: bigfloat(3.14159265359)
    """
    return literal(lt.bigfloat(value))


def bigint(value: int) -> Term:
    """Create a bigint literal.

    Example: bigint(9223372036854775808)
    """
    return literal(lt.bigint(value))


def binary(value: bytes) -> Term:
    """Create a binary data literal from bytes.

    Example: binary(b'\\x48\\x65\\x00\\xff\\x20\\x7a\\x1b\\x80')
    """
    return literal(lt.binary(value))


def boolean(value: bool) -> Term:
    """Create a boolean literal.

    Example: boolean(True)
    """
    return literal(lt.boolean(value))


def char(value: str) -> Term:
    """Create a character term (represented as int32 code point).

    Example: char('A')  # Creates int32(65)
    """
    return int32(ord(value))


def comparison(comp: hydra.util.Comparison) -> Term:
    """Create a comparison result term (lessThan, equalTo, or greaterThan)."""
    match comp:
        case hydra.util.Comparison.EQUAL_TO:
            return inject_unit(
                hydra.util.COMPARISON__NAME,
                hydra.util.COMPARISON__EQUAL_TO__NAME
            )
        case hydra.util.Comparison.LESS_THAN:
            return inject_unit(
                hydra.util.COMPARISON__NAME,
                hydra.util.COMPARISON__LESS_THAN__NAME
            )
        case hydra.util.Comparison.GREATER_THAN:
            return inject_unit(
                hydra.util.COMPARISON__NAME,
                hydra.util.COMPARISON__GREATER_THAN__NAME
            )


def compose(f: Term, g: Term) -> Term:
    """Compose two functions (apply g then f) to create a new function.

    Example: compose(var("stringLength"), var("toString"))
    This creates a function equivalent to: lambda x: stringLength(toString(x))
    Function composition applies right-to-left: (f . g)(x) = f(g(x))
    """
    return TermFunction(
        FunctionLambda(Lambda(Name("arg_"), Nothing(), apply(f, apply(g, var("arg_")))))
    )


def constant(term: Term) -> Term:
    """Create a constant function that always returns the same value.

    Example: constant(true())  # A function that always returns True
    """
    return lambda_(hydra.constants.ignored_variable, term)


def elimination(elim: Elimination) -> Term:
    """Construct a term from an elimination (record projection, union case, or unwrap)."""
    return TermFunction(FunctionElimination(elim))


def false() -> Term:
    """Boolean false literal."""
    return boolean(False)


def field(n: str, t: Term) -> Field:
    """Create a field with the given name and value.

    Example: field("age", int32(30))
    """
    return Field(Name(n), t)


def fields_to_map(fields: Sequence[Field]) -> FrozenDict[Name, Term]:
    """Convert a sequence of fields to a map from names to terms."""
    return FrozenDict({field.name: field.term for field in fields})


def first() -> Term:
    """First element projection function for pairs."""
    return primitive(Name("hydra.lib.pairs.first"))


def float32(value: float) -> Term:
    """Create a float32 literal.

    Example: float32(3.14)
    """
    return literal(lt.float32(value))


def float64(value: float) -> Term:
    """Create a float64 literal.

    Example: float64(3.14159265359)
    """
    return literal(lt.float64(value))


def float_(value: FloatValue) -> Term:
    """Create a floating-point literal with specified precision.

    Example: float_(FloatValueFloat32(3.14))
    """
    return literal(lt.float_(value))


def identity() -> Term:
    """Identity function that returns its argument unchanged."""
    return lambda_("x_", var("x_"))


def inject(tname: Name, fname: Name, term: Term) -> Term:
    """Create a union value by injecting a value into a specific variant.

    Example: inject(Name("Result"), Name("success"), int32(42))
    This creates a "Result" union with the "success" variant containing value 42.
    Use this to construct values of union types at runtime.
    """
    return TermUnion(Injection(tname, Field(fname, term)))


def int16(value: int) -> Term:
    """Create an int16 literal.

    Example: int16(32767)
    """
    return literal(lt.int16(value))


def int32(value: int) -> Term:
    """Create an int32 literal.

    Example: int32(42)
    """
    return literal(lt.int32(value))


def int64(value: int) -> Term:
    """Create an int64 literal.

    Example: int64(9223372036854775807)
    """
    return literal(lt.int64(value))


def int8(value: int) -> Term:
    """Create an int8 literal.

    Example: int8(127)
    """
    return literal(lt.int8(value))


def integer(value: IntegerValue) -> Term:
    """Create an integer literal with specified bit width.

    Example: integer(IntegerValueInt32(42))
    """
    return literal(lt.integer(value))


def just(term: Term) -> Term:
    """Create a 'Just' optional value.

    Example: just(string("found"))
    """
    return optional(Just(term))


def lambda_(param: str, body: Term) -> Term:
    """Create a lambda function with one parameter.

    Example: lambda_("x", apply(var("x"), int32(1)))
    """
    return TermFunction(FunctionLambda(Lambda(Name(param), Nothing(), body)))


def lambdas(params: Sequence[str], body: Term) -> Term:
    """Create a multi-parameter lambda function (curried).

    Example: lambdas(["x", "y"], apply_all(var("add"), [var("x"), var("y")]))
    This creates the function: lambda x: lambda y: add(x, y)
    """
    if not params:
        return body
    return lambda_(params[0], lambdas(params[1:], body))


def lambda_typed(param: str, dom: Type, body: Term) -> Term:
    """Create a lambda function with an explicit domain type.

    Example: lambda_typed("x", Types.int32, list_([var("x")]))
    """
    return TermFunction(FunctionLambda(Lambda(Name(param), Just(dom), body)))


def let_multi(bindings: Sequence[tuple[str, Term]], body: Term) -> Term:
    """Create a let expression with multiple bindings.

    Example: let_multi([("x", int32(1)), ("y", int32(2))], apply_all(var("add"), [var("x"), var("y")]))
    """
    return TermLet(
        Let(
            tuple([Binding(Name(name), term, Nothing()) for name, term in bindings]),
            body,
        )
    )


def let_term(v: Name, t1: Term, t2: Term) -> Term:
    """Create a let expression with a single binding."""
    return TermLet(Let(tuple([Binding(v, t1, Nothing())]), t2))


def lets(bindings: Sequence[Field], env: Term) -> Term:
    """Create a let term with any number of bindings.

    Example: lets([field("x", int32(1)), field("y", int32(2))], pair(var("x"), var("y")))
    """
    def to_binding(f: Field) -> Binding:
        return Binding(f.name, f.term, Nothing())

    return TermLet(Let(tuple([to_binding(f) for f in bindings]), env))


def lets_typed(bindings: Sequence[tuple[str, Term, TypeScheme]], env: Term) -> Term:
    """Create a let expression with typed bindings."""
    def to_binding(b: tuple[str, Term, TypeScheme]) -> Binding:
        name, term, ts = b
        return Binding(Name(name), term, Just(ts))

    return TermLet(Let(tuple([to_binding(b) for b in bindings]), env))


def left(term: Term) -> Term:
    """Create a 'Left' either value.

    Example: left(string("error"))
    """
    return TermEither(Left(term))


def list_(terms: Sequence[Term]) -> Term:
    """Create a list of terms.

    Example: list_([int32(1), int32(2), int32(3)])
    """
    return TermList(tuple(terms))


def literal(value: Literal) -> Term:
    """Create a term from a literal value.

    Example: literal(LiteralString("hello"))
    """
    return TermLiteral(value)


def map_(terms: Mapping[Term, Term]) -> Term:
    """Create a map/dictionary term.

    Example: map_({string("January"): int32(31), string("February"): int32(28)})
    """
    return TermMap(FrozenDict(terms))


def map_term(terms: Mapping[Term, Term]) -> Term:
    """Create a map/dictionary term (alias for map_)."""
    return TermMap(FrozenDict(terms))


def match(tname: Name, def_: Maybe[Term], fields: Sequence[Field]) -> Term:
    """Create a pattern match on a union type.

    Example: match(Name("Result"), Just(string("unknown")),
                   [field("success", lambda_("s", apply(var("processSuccess"), var("s")))),
                    field("error", lambda_("e", apply(var("handleError"), var("e"))))])
    This allows handling different cases of a union type with specific logic for each variant.
    The optional second parameter provides a default case for any unmatched variants.
    """
    return TermFunction(
        FunctionElimination(EliminationUnion(CaseStatement(tname, def_, tuple(fields))))
    )


def match_with_variants(
        tname: Name, def_: Maybe[Term], pairs: Sequence[tuple[Name, Name]]
) -> Term:
    """Create a match term that maps variant names to other variants."""
    return match(
        tname,
        def_,
        [Field(from_, constant(inject_unit(tname, to_))) for from_, to_ in pairs],
    )


def nothing() -> Term:
    """Create a 'Nothing' optional value."""
    return optional(Nothing())


def optional(term: Maybe[Term]) -> Term:
    """Create an optional (nullable) term.

    Example: optional(Just(string("found")))
    """
    return TermMaybe(term)


def pair(a: Term, b: Term) -> Term:
    """Create a pair.

    Example: pair(string("name"), int32(42))
    """
    return TermPair((a, b))


def primitive(name: Name) -> Term:
    """Create a primitive function reference.

    Example: primitive(Name("hydra.lib.strings.length"))
    """
    return TermFunction(FunctionPrimitive(name))


def project(tname: Name, fname: Name) -> Term:
    """Create a field projection function.

    Example: project(Name("Person"), Name("firstName"))
    """
    return TermFunction(
        FunctionElimination(EliminationRecord(Projection(tname, fname)))
    )


def record(tname: Name, fields: Sequence[Field]) -> Term:
    """Create a record with named fields of the specified type.

    Example: record(Name("Person"), [
               field("name", string("John")),
               field("age", int32(30)),
               field("email", string("john@example.com"))])
    Records are products of named fields with values that can be accessed by field name.
    """
    return TermRecord(Record(tname, tuple(fields)))


def right(term: Term) -> Term:
    """Create a 'Right' either value.

    Example: right(int32(42))
    """
    return TermEither(Right(term))


def second() -> Term:
    """Second element projection function for pairs."""
    return primitive(Name("hydra.lib.pairs.second"))


def set_(s: set[Term]) -> Term:
    """Create a set of terms.

    Example: set_({string("a"), string("b"), string("c")})
    """
    return TermSet(frozenset(s))


def string(value: str) -> Term:
    """Create a string literal.

    Example: string("hello world")
    """
    return literal(lt.string(value))


def triple(a: Term, b: Term, c: Term) -> Term:
    """Create a triple using nested pairs."""
    return pair(a, pair(b, c))


def true() -> Term:
    """Boolean true literal."""
    return boolean(True)


def tuple_(terms: Sequence[Term]) -> Term:
    """Create a tuple using nested pairs.

    Example: tuple_([a, b, c]) creates pair(a, pair(b, c))
    """
    if len(terms) == 0:
        return unit()
    elif len(terms) == 1:
        return terms[0]
    elif len(terms) == 2:
        return pair(terms[0], terms[1])
    else:
        return pair(terms[0], tuple_(terms[1:]))


def tuple2(a: Term, b: Term) -> Term:
    """Create a 2-tuple (same as pair)."""
    return pair(a, b)


def tuple3(a: Term, b: Term, c: Term) -> Term:
    """Create a 3-tuple using nested pairs."""
    return pair(a, pair(b, c))


def tuple4(a: Term, b: Term, c: Term, d: Term) -> Term:
    """Create a 4-tuple using nested pairs."""
    return pair(a, pair(b, pair(c, d)))


def tuple5(a: Term, b: Term, c: Term, d: Term, e: Term) -> Term:
    """Create a 5-tuple using nested pairs."""
    return pair(a, pair(b, pair(c, pair(d, e))))


def type_application(term: Term, types: Sequence[Type]) -> Term:
    """Apply type arguments to a polymorphic term.

    Example: type_application(var("map"), [Types.int32, Types.string])
    This instantiates a polymorphic function with concrete types.
    For instance, if 'map' has type 'forall a b. (a -> b) -> list a -> list b',
    the example would instantiate it to '(int32 -> string) -> list int32 -> list string'.
    """
    return reduce(
        lambda t, ty: TermTypeApplication(TypeApplicationTerm(t, ty)),
        types,
        term
    )


def type_lambda(vars: Sequence[Name], body: Term) -> Term:
    """Create a type abstraction (universal quantification).

    Example: type_lambda([Name("a"), Name("b")],
               lambda_typed("f", Types.function(Types.var("a"), Types.var("b")),
                 lambda_typed("x", Types.var("a"), apply(var("f"), var("x")))))
    This creates a polymorphic term with type variables.
    The example creates a higher-order function with type 'forall a b. (a -> b) -> a -> b'.
    """
    return reduce(
        lambda b, v: TermTypeLambda(TypeLambda(v, b)),
        vars,
        body
    )


def tyapp(term: Term, typ: Type) -> Term:
    """Apply a single type argument to a polymorphic term."""
    return TermTypeApplication(TypeApplicationTerm(term, typ))


def tyapps(term: Term, types: Sequence[Type]) -> Term:
    """Apply type arguments to a polymorphic term (alias for type_application)."""
    return type_application(term, types)


def tylam(var: str, body: Term) -> Term:
    """Create a type abstraction with a single type variable."""
    return TermTypeLambda(TypeLambda(Name(var), body))


def tylams(vars: Sequence[str], body: Term) -> Term:
    """Create a type abstraction with multiple type variables."""
    return reduce(
        lambda b, v: TermTypeLambda(TypeLambda(Name(v), b)),
        reversed(vars),
        body
    )


def uint16(value: int) -> Term:
    """Create a uint16 literal.

    Example: uint16(65535)
    """
    return literal(lt.uint16(value))


def uint32(value: int) -> Term:
    """Create a uint32 literal.

    Example: uint32(4294967295)
    """
    return literal(lt.uint32(value))


def uint64(value: int) -> Term:
    """Create a uint64 literal.

    Example: uint64(18446744073709551615)
    """
    return literal(lt.uint64(value))


def uint8(value: int) -> Term:
    """Create a uint8 literal.

    Example: uint8(255)
    """
    return literal(lt.uint8(value))


def unit() -> Term:
    """Unit value (empty record)."""
    return TermUnit()


def inject_unit(tname: Name, fname: Name) -> Term:
    """Create a unit variant of a union (convenience function).

    Example: inject_unit(Name("Result"), Name("success"))
    Equivalent to inject but automatically uses unit as the value.
    """
    return inject(tname, fname, unit())




def unwrap(name: Name) -> Term:
    """Create an unwrap function for a wrapped type.

    Example: unwrap(Name("Email"))
    """
    return TermFunction(FunctionElimination(EliminationWrap(name)))


def var(name: str) -> Term:
    """Create a variable reference.

    Example: var("x")
    """
    return TermVariable(Name(name))


def with_variant(tname: Name, fname: Name) -> Term:
    """Create a constant function that produces a unit variant."""
    return constant(inject_unit(tname, fname))


def wrap(name: Name, term: Term) -> Term:
    """Create a wrapped term (instance of a newtype).

    Example: wrap(Name("Email"), string("user@example.com"))
    """
    return TermWrap(WrappedTerm(name, term))
